README.md
# primality.js

Primality is a JavaScript library for prime numbers. It features a fantastic
[primality test](https://en.wikipedia.org/wiki/Primality_test) and
identification of the various
[classes](https://en.wikipedia.org/wiki/Template:Prime_number_classes) of prime
numbers.

## Features

  - Check the primality of a number
  - Works with numbers disguised as strings
  - Also does arrays
  - Provides checks for the following classes of prime numbers:
    - [twin](https://en.wikipedia.org/wiki/Twin_prime)
    - [cousin](https://en.wikipedia.org/wiki/Cousin_prime)
    - [sexy](https://en.wikipedia.org/wiki/Sexy_prime)
    - [Wilson](https://en.wikipedia.org/wiki/Wilson_prime)
  - About 1 kibibyte minified and gzipped
  - ESM support

## Installation

``` bash
$ npm install primality
```

## Usage

Primality's flagship method is `primality()`, which works as you might expect it
to:

``` javascript
import primality from 'primality';

primality(7);
// => true

primality(6);
// => false
```

Of course, you can pass strings instead of numbers if you'd like:

``` javascript
primality('13');
// => true
```

Primality can even do arrays. If any of the values in an array are not prime,
`false` is returned.

``` javascript
primality([17, 19, 23]);
// => true

primality([17, 20, 23]);
// => false
```

Beyond primality testing, Primality can also tell you if a pair of numbers are
twin, cousin, or sexy primes. Twin primes are prime numbers that differ from
each other by two. Similarly, cousin primes differ by four and sexy primes
differ by six.

``` javascript
import { areCousinPrimes, areSexyPrimes, areTwinPrimes } from 'primality';

areCousinPrimes(3, 7);
// => true

areSexyPrimes(5, 11);
// => true

areTwinPrimes(3, 5);
// => true
```

You can also check for Wilson primes. Only three Wilson primes are known at the
moment: 5, 13, and 563.

``` javascript
import { isWilsonPrime } from 'primality';

isWilsonPrime(563);
// => true
```


CHANGELOG.md
## [3.0.0](https://github.com/kenany/primality/compare/2.0.0...3.0.0) (2023-09-07)


### ⚠ BREAKING CHANGES

* Node.js v14 and v16 are no longer supported.

### Features

* drop Node.js v14 and v16 ([186de9e](https://github.com/kenany/primality/commit/186de9e27a546100a369b1eb06b9c15f443ffd07))

## [2.0.0](https://github.com/kenany/primality/compare/1.6.1...2.0.0) (2023-04-22)


### ⚠ BREAKING CHANGES

* This is now a pure ESM package. Additionally, support
for bower and component has been dropped.
* Node.js <14.14.0 is no longer supported.

### Features

* overhaul for ESM ([c81fbae](https://github.com/kenany/primality/commit/c81fbae36946665d5a01771adb40d6373d2b43c7))
* require at least Node.js v14.14.0 ([09283a5](https://github.com/kenany/primality/commit/09283a5cbd4def0c600618362cdb76d8b10f716c))


### Bug Fixes

* **deps:** lodash.isfinite@2.4.1 ([fb7712e](https://github.com/kenany/primality/commit/fb7712eab502abc1c27aea28b0cf8fc815d7c0f4))
* update copyright year ([138b7b2](https://github.com/kenany/primality/commit/138b7b24e0d0a2519b0931b18a63e5176055564f))


primality.js
import factorial from 'factorial';
import isFinite from 'lodash.isfinite';

const WILSON_PRIMES = [5, 13, 563];
const WIEFERICH_PRIMES = [1093, 3511];

/**
 * Finds the smallest factor of `n`
 *
 * @private
 * @param {number} n The value to check
 * @returns {number}
 *   The smallest prime that divides n
 *   NaN if n is NaN or Infinity
 *   0 if n is 0
 *   1 if n = 1, n = -1, or n is not an integer
 */
function leastFactor(n) {
  if (n === 0) return 0;
  else if (n % 1 || n * n < 2) return 1;
  else if (n % 2 === 0) return 2;
  else if (n % 3 === 0) return 3;
  else if (n % 5 === 0) return 5;

  const m = Math.sqrt(n);
  for (let i = 7; i <= m; i += 30) {
    if (n % i === 0) return i;
    else if (n % (i + 4) === 0) return i + 4;
    else if (n % (i + 6) === 0) return i + 6;
    else if (n % (i + 10) === 0) return i + 10;
    else if (n % (i + 12) === 0) return i + 12;
    else if (n % (i + 16) === 0) return i + 16;
    else if (n % (i + 22) === 0) return i + 22;
    else if (n % (i + 24) === 0) return i + 24;
  }
  return n;
}

/**
 * Checks if `value` is prime.
 *
 * @private
 * @param {number | string} value The value to check
 * @returns {boolean} Returns `true` if `value` is prime
 */
function isPrime(value) {
  if (Number.isNaN(value) || !isFinite(value) || value % 1 || value < 2) {
    return false;
  }
  if (value !== leastFactor(value)) return false;
  return true;
}

/**
 * Primality test.
 *
 * @param {number | string | readonly (number | string)[]} input A number,
 *  string, or array to check the primality of.
 * @returns {Boolean} Returns `true` if `input` is prime.
 * @example
 *
 * primality(7);
 * // => true
 *
 * primality('13');
 * // => true
 *
 * primality([17, 19, 23]);
 * // => true
 */
export default function primality(input) {
  if (input === null || input === '') return null;
  else if (Array.isArray(input)) {
    let i = input.length;
    while (i--) {
      if (!isPrime(input[i])) return false;
    }
    return true;
  }
  else return isPrime(input);
}

/**
 * Checks if `a` and `b` are primes which differ by `difference`.
 *
 * @private
 * @param {Number} a First of the pair
 * @param {Number} b Second of the pair
 * @param {Number} difference
 * @returns {Boolean}
 */
function isRelated(a, b, difference) {
  return Math.abs(a - b) !== difference
    ? false
    : primality([a, b]);
}
/**
 * Checks if `a` and `b` are twin primes
 *
 * <https://en.wikipedia.org/wiki/Twin_prime>
 *
 * @param {Number} a First of the pair
 * @param {Number} b Second of the pair
 * @returns {Boolean} Returns `true` if `a` and `b` are twin primes
 * @example
 *
 * primality.areTwinPrimes(3, 5)
 * // => true
 */
export function areTwinPrimes(a, b) {
  return isRelated(a, b, 2);
}

/**
 * Checks if `a` and `b` are cousin primes
 *
 * <https://en.wikipedia.org/wiki/Cousin_prime>
 *
 * @param {Number} a First of the pair
 * @param {Number} b Second of the pair
 * @returns {Boolean} Returns `true` if `a` and `b` are cousin primes
 * @example
 *
 * primality.areCousinPrimes(3, 7)
 * // => true
 */
export function areCousinPrimes(a, b) {
  return isRelated(a, b, 4);
}

/**
 * Checks if `a` and `b` are sexy primes
 *
 * <https://en.wikipedia.org/wiki/Sexy_prime>
 *
 * @param {Number} a First of the pair
 * @param {Number} b Second of the pair
 * @returns {Boolean} Returns `true` if `a` and `b` are sexy primes
 * @example
 *
 * primality.areSexyPrimes(5, 11)
 * // => true
 */
export function areSexyPrimes(a, b) {
  return isRelated(a, b, 6);
}

/**
 * Checks if `value` is a Wilson prime.
 *
 * <https://en.wikipedia.org/wiki/Wilson_prime>
 *
 * @param {Number} value
 * @returns {Boolean} Returns `true` if `value` is a Wilson prime.
 * @example
 *
 * primality.isWilsonPrime(5);
 * // => true
 */
export function isWilsonPrime(value) {
  return WILSON_PRIMES.includes(value)
    || (factorial(value - 1, { useBigInt: true }) + 1n)
      % BigInt(Math.pow(value, 2)) === 0;
}

/**
 * Checks if `value` is a Wieferich prime.
 *
 * <https://en.wikipedia.org/wiki/Wieferich_prime>
 *
 * @param {Number} value
 * @returns {Boolean} Returns `true` if `value` is a Wieferich prime.
 * @example
 *
 * primality.isWieferichPrime(1093);
 * // => true
 */
export function isWieferichPrime(value) {
  return WIEFERICH_PRIMES.includes(value)
    || (Math.pow(2, value - 1) - 1) % Math.pow(value, 2) === 0;
}
